File: c_introduction/cash/cash.c
#include <cs50.h>
#include <stdio.h>

int calculate_dimes(int cents);
int calculate_nickels(int cents);
int calculate_pennies(int cents);
int calculate_quarters(int cents);

// Magic Numbers

#define DIME_VALUE 10
#define NICKEL_VALUE 5
#define PENNIE_VALUE 1
#define QUARTER_VALUE 25

// Greedy Algorithms for quarters (25¢), dimes (10¢), nickels (5¢), and pennies (1¢)

int main(void)
{
    // Promp user for change owned, unit: cents
    int cents;
    do
    {
        cents = get_int("Change owed: ");
    }
    while (cents < 0);

    // Calculating quarters and subtracting that value fron int cent
    int quarters = calculate_quarters(cents);
    cents -= quarters * QUARTER_VALUE;

    // Calculating dimes and subtracting that value fron int cent
    int dimes = calculate_dimes(cents);
    cents -= dimes * DIME_VALUE;

    // Calculating nickels and subtracting that value fron int cent
    int nickels = calculate_nickels(cents);
    cents -= nickels * NICKEL_VALUE;

    // Calculating pennies and subtracting that value is not needed
    int pennies = calculate_pennies(cents);

    int sum = quarters + dimes + nickels + pennies;
    printf("%i\n", sum);
}

int calculate_quarters(int cents)
{
    int coins = 0;
    int coin_value = QUARTER_VALUE;
    while (cents >= coin_value)
    {
        coins++;
        cents -= coin_value;
    }
    return coins;
}

int calculate_dimes(int cents)
{
    int coins = 0;
    int coin_value = DIME_VALUE;
    while (cents >= coin_value)
    {
        coins++;
        cents -= coin_value;
    }
    return coins;
}

int calculate_nickels(int cents)
{
    int coins = 0;
    int coin_value = NICKEL_VALUE;
    while (cents >= coin_value)
    {
        coins++;
        cents -= coin_value;
    }
    return coins;
}

int calculate_pennies(int cents)
{
    int coins = 0;
    int coin_value = PENNIE_VALUE;
    while (cents >= coin_value)
    {
        coins++;
        cents -= coin_value;
    }
    return coins;
}

File: python/rewrite_c_in_python_cash/cash.py
from cs50 import get_float

# Coin values
dime = 10
nickel = 5
pennie = 1
quarter = 25

# Promp user for change owned, unit: dollars
while True:
    dollars = get_float("Change in dollars, please: ")
    if dollars > 0:
        break


def main():

    # Converting dollars to cents, floats to int
    cents = int(dollars * 100)

    # Calculating quarter coins and subtracting that value fron int cent
    quarters = calculate_quarters(cents)
    remaining_cents = cents - quarters * quarter

    # Calculating dime coins and subtracting that value fron int cent
    dimes = calculate_dimes(remaining_cents)
    remaining_cents -= dimes * dime

    # Calculating nickel coins and subtracting that value fron int cent
    nickels = calculate_nickels(remaining_cents)
    remaining_cents -= nickels * nickel

    # Calculating pennie coins and subtracting that value fron int cent
    pennis = calculate_pennis(remaining_cents)

    # Summing up all coins and printing the sum
    coins = quarters + dimes + nickels + pennis
    print(coins)


def calculate_quarters(n):
    coins = 0
    while n >= quarter:
        coins += 1
        n -= quarter
    return coins


def calculate_dimes(n):
    coins = 0
    while n >= dime:
        coins += 1
        n -= dime
    return coins


def calculate_nickels(n):
    coins = 0
    while n >= nickel:
        coins += 1
        n -= nickel
    return coins


def calculate_pennis(n):
    coins = 0
    while n >= pennie:
        coins += 1
        n -= pennie
    return coins


main()
